File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/iterator/value/IteratorValueOf2On2.java
package joookiwi.collection.java.iterator.value;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Range;

import static joookiwi.collection.java.CommonContracts.ALWAYS_FALSE_0;
import static joookiwi.collection.java.CommonContracts.ALWAYS_TRUE_0;

/// An implementation of an [InsideIteratorValue] with only the 2<sup>nd</sup> value out of 2
/// if a [CollectionIterator][joookiwi.collection.java.iterator.CollectionIterator]
///
/// @param <T> The element type
@NotNullByDefault
public class IteratorValueOf2On2<T extends @Nullable Object>
        extends AbstractIteratorValue<T>
        implements InsideIteratorValue<T> {

    private final T __value;

    public IteratorValueOf2On2(final T value) {
        super();
        __value = value;
    }

    @Contract(pure = true)
    @Override public @Range(from = 1, to = 1) int index() { return 1; }

    @Contract(value = ALWAYS_FALSE_0, pure = true)
    @Override public boolean isDone() { return false; }

    @Contract(value = ALWAYS_TRUE_0, pure = true)
    @Override public boolean isNotDone() { return true; }

    @Contract(pure = true)
    @Override public T value() { return __value; }

}

File: javascript/src/iterator/value/IteratorValueOf2On2.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {InsideIteratorValue} from "./InsideIteratorValue"

import {AbstractIteratorValue} from "./AbstractIteratorValue"

/**
 * An implementation of an {@link InsideIteratorValue} with only the second value out of two
 * of a {@link CollectionIterator}
 *
 * @typeParam T The element type (`unknown` by default)
 */
export class IteratorValueOf2On2<const T = unknown, >
    extends AbstractIteratorValue<T, false>
    implements InsideIteratorValue<T> {

    readonly #value

    public constructor(value: T,) {
        super()
        this.#value = value
    }

    public get index(): 1 { return 1 }

    public override isDone(): false { return false }

    public override isNotDone(): true { return true }

    public override get value(): T { return this.#value }

}
